### PR TITLE
feat: add clip_hg executor

### PR DESCRIPTION
**Pull request by ZiniuYu**

### CODE REVIEW

File: server/clip_server/executors/clip_hg.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        )
+        image_docs = DocumentArray(
+            filter(
+                lambda x: (
+                    (x.tensor is not None or x.blob != b'' or x.uri)
+                    and (overwrite_embeddings or x.embedding is None)
+                ),
+                docs[traversal_paths],
+            )
+        )
```
Comment (numb3r3) [inline server/clip_server/executors/clip_hg.py:134]: Polish this block by following implementations in other backends.

File: server/clip_server/executors/clip_hg.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+        with torch.inference_mode():
+            for batch in text_docs.batch(batch_size=batch_size):
+                self._encode_texts(batch)
+
+            for batch in image_docs.batch(batch_size=batch_size):
+                self._encode_images(batch)
```
Comment (numb3r3) [inline server/clip_server/executors/clip_hg.py:141]: Please also apply the optimization tricks introduced in other backends.

File: server/clip_server/executors/clip_hg.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            tensors_batch.append(tensor)
+        if self.use_default_preprocessing:
+            tensor = self._preprocess_images(tensors_batch)
+        else:
+            tensor = {
+                'pixel_values': torch.tensor(
+                    batch.tensors, dtype=torch.float32, device=self.device
+                )
+            }
+
```
Comment (numb3r3) [inline server/clip_server/executors/clip_hg.py:167]: separate `preprocess` and `encode` steps into different sub-functions so as to get better monitoring, refer to this https://github.com/jina-ai/clip-as-service/pull/738

**Review by numb3r3 — CHANGES_REQUESTED** [review]
- add `rank()` endpoint
- add unittest
- add dependency of `transformers` into `setup.py`

**Review by hanxiao — CHANGES_REQUESTED** [review]
what's the point of this PR? is it just for practicing how to implement executor?

**Comment by numb3r3** [comment]
@hanxiao This PR is mainly developed for the community. If the users who already use (or finetune) the clip model from huggingface, then this PR can enable them to easily enjoy the power of jina.

File: server/setup.py [hunk 4]
```diff
@@ -60,6 +59,7 @@
         ]
         + (['onnxruntime-gpu>=1.8.0'] if sys.platform != 'darwin' else []),
         'tensorrt': ['nvidia-tensorrt'],
+        'hg': ['transformers>=4.16.2'],
```
Comment (numb3r3) [inline server/setup.py:62]: use `huggingface` instead of `hg`

Comment (numb3r3) [reply server/setup.py:62]: And the commit message should not be `chore:`

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/executors/clip_hg.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
+import os
+import warnings
+from multiprocessing.pool import ThreadPool
+from typing import Dict, Optional, Sequence
+import numpy as np
+import torch
+from copy import deepcopy
```
Comment (numb3r3) [inline server/clip_server/executors/clip_hg.py:7]: remove this import if the `copy` is deleted

Comment (numb3r3) [reply server/clip_server/executors/clip_hg.py:7]: and clean the unused imports

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Pull request merged**